### ISSUE TITLE
[BUG] It is not possible to calculate RMSSE (mean_squared_scaled_error() function) and MASE (mean_absolute_scaled_error() function) metrics for time series of different lengths.

### ISSUE THREAD

**Issue by Richdadua** [sktime/sktime#5102]
**Describe the bug**
<--
 In real life, when forecasting a large number of time series, we are faced with very different time series in length. When I made a wide format dataset in which each time series became a separate column, then in the y_train dataset in shorter time series, earlier dates were filled with NaN values. I thought that there is nothing wrong with this, and this should not interfere with calculating the MSE of the naive algorithm for each time serie. However, the function "broke", swearing at the presence of NaN.
I figured out that the error is intentionally thrown by sklearn, which "requires" time series of the same length. This is correctly done for the test period. But in the metrics under consideration, it is necessary to calculate the errors of the naive algorithm on training periods, and this requirement of the same lengths of time series harms us. Therefore, in these functions, the calculation of the errors of the naive algorithm does not need to be sent to the calculation in sklearn. 
-->

**To Reproduce**
<--
Below is a simple example, with two time series, one of which is one observation longer in the historical period.
-->

```python
from sktime.performance_metrics.forecasting import mean_squared_scaled_error
from sktime.performance_metrics.forecasting import mean_absolute_scaled_error
y_train = np.array([[1.5, np.nan], [0.5, 1], [-1, 1], [7, -6]])
y_true = np.array([[0.5, 1], [-1, 1], [7, -6]])
y_pred = np.array([[0, 2], [-1, 2], [8, -5]])
mean_squared_scaled_error(y_true, y_pred, y_train=y_train, square_root=True)
# ValueError: Input contains NaN.
mean_absolute_scaled_error(y_true, y_pred, y_train=y_train)
# ValueError: Input contains NaN.
```

**Expected behavior**
<--
The calculation must be performed without issuing an error.
-->

**Additional context**

**Versions**
Versions
0.21.0

System:
python: 3.9.16 | packaged by conda-forge | (main, Feb 1 2023, 21:28:38) [MSC v.1929 64 bit (AMD64)]
executable: C:\Users...\miniconda3\envs\ml_ts\python.exe
machine: Windows-10-10.0.19045-SP0

Python dependencies:
pip: 23.1.1
sktime: 0.21.0
sklearn: 1.2.2
skbase: 0.4.6
numpy: 1.23.5
scipy: 1.10.1
pandas: 2.0.0
matplotlib: 3.7.1
joblib: 1.2.0
statsmodels: 0.13.5
numba: 0.56.4
pmdarima: 2.0.2
tsfresh: 0.0.post0.dev58+ga42b9ea.dirty
tensorflow: None
tensorflow_probability: None

**Comment by fkiraly** [comment]
reproduced on current `main`, python 3.11, windows

**Comment by yarnabrina** [comment]
If it was the case of a hierarchical dataset with different series are available at different horizons, like Jan-Dec, Jan-Jun and Jul-Dec, and `sktime` fails in that case, I'd readily agree that is a framework bug. But in this particaular example, I'm a bit concerned.

@Richdadua, if you don't mind me asking, can you please explain why do you think it is a bug? What is the expected value of RMSSE or MASE for the provided example?

---

Question to @fkiraly: how does `sktime` calculate metrics for multiple or multivariate cases? Does it calculate per instance/variable/both and then return raw/average based on `multioutput`, or is it based on a flattened 1D array? My first look suggested the latter, which seemed a bit questionable, so confirming.

**Comment by fkiraly** [comment]
> how does `sktime` calculate metrics for multiple or multivariate cases?

I think it is inconsistent and needs to be improved, but there's a majority case.

In the majority case, the metric is purely univariate and single-instance.

In this case, the metric has a `multioutput` and a `multilevel` argument, which controls the averaging over the variables resp instances.

However, there are metrics which are genuinely multivariate or multi-instance, and the framework (or tests) do not capture such behaviour.
Examples for genuinely multivariate metrics are multivariate log-loss, or whitening-plus-metric.

Therefore, I think there should be a tag that tells the user and the framework whether a metric is genuinely multivariate or multi-instance. This would imply whether you expect a `multioutput` argument, say, and whether it should behave in a certain way.

**Comment by yarnabrina** [comment]
> In this case, the metric has a multioutput and a multilevel argument, which controls the averaging over the variables resp instances.

I was looking only at `sktime/performance_metrics/forecasting/_functions.py`, and they miss this. So, I realise I have been misusing the metrics so far in my use cases, and should have used the classes. THanks for pointing this out, it was not obvious to me.

But I am still confused, does it also not mean that functions are always non-hierarchical, and the reported code example is not a bug for `mean_absolute_scaled_error`? May be it's a bug for `MeanAbsoluteScaledError`, but not for the function variant which is not expected to support it by design?

> Another example is the current implementation of RMSSE, which is a fraction of averages rather than the average of fractions - where @Richdadua was expecting the latter and not the former; the latter is not genuinely multivariate and the "natural" multivariate generalization of RMSSE-univariate.

Is it coming from some other context? I think I'm missing that, so please ignore my above remark about not-being-bug.

**Comment by Richdadua** [comment]
Colleagues, I'm a little confused. I want to clarify the situation with the terms. Here are my theses:
1) If hierarchical forecasting occurs, then the Weighted RMSSE must be calculated at each level of the hierarchy separately. That is, do not mix time series from different hierarchies into one general metric.
2) To calculate the Weighted RMSSE for all time series of the same hierarchical level and there is a multioutput parameter. When it is equal to "uniform_average", then the weights of all time series are equal. And when a vector of weights is supplied, the calculation proceeds according to this vector.
3) @fkiraly you write the terms "multilevel", "multivariate", "multi-instance". I think you're overcomplicating the problem. I reduce the problem to the following: "We have many time series of the same origin that we predicted. For each of them we calculated some kind of metric (in our case RMSSE). And I now want to get one general number of this metric, according to the calculated weights (or without them - then they are equal for all time series).”

**Comment by fkiraly** [comment]
@Richdadua, agreed on 1.

Re 2 and 3, it is important to note that there are two dimensions we could be averaging over: time series instance index, and variable index. `multioutput` controls the averaging over variables, `multilevel` controls the averaging over time series instance index.

From a formal/logic standpoint, `multilevel` and `multioutput` seem required to parameterize the potential intentions of how we report or reconcile the forecasts across variables, and across instances.

**Comment by fkiraly** [comment]
Regarding multiple instances, I think this works?

```python
import pandas as pd
from sktime.performance_metrics.forecasting import MeanSquaredScaledError

y_train = pd.DataFrame([1.5, 0.5, -1, 7, 0.5, 1, -6], index=pd.MultiIndex.from_tuples([(0, 0), (0, 1), (0, 2), (0, 3), (1, 1), (1, 2), (1, 3)]))
y_test = pd.DataFrame([0.5, -1, 7, 1, 1, -6], index=pd.MultiIndex.from_product([[0, 1], [4, 5, 6]]))
y_pred = pd.DataFrame([0, -1, 8, 2, 2, -5], index=pd.MultiIndex.from_product([[0, 1], [4, 5, 6]]))

msse = MeanSquaredScaledError(square_root=True)
msse(y_test, y_pred, y_train=y_train)
```

This gives 0.169... as in your example.

**Comment by fkiraly** [comment]
PS: you could also pass the `y_train` etc as a 3D numpy array in the format closest to the math formalism I introduced.

**Comment by Richdadua** [comment]
Yes, it looks like the option of calculating the metrics through a class instance works correctly. 1) It handles time series of different lengths. 2) It calculates the (R)MSSE metrics of each time series and then applies weights.
I recommend everyone to use the class implementation.
